### PR TITLE
TW-221: Fix remove url markdown on messages

### LINKED ISSUE #221: [Chat] A part of url is not underline in message bubble
OS:  Android
Device: Mobile
TMail version:  2.1.3

*Precondition: 
- User A send user B a link which having thumbnail.

*Data:
- Example link: 
https://dantri.com.vn/xa-hoi/xu-ly-2-dia-diem-co-hien-tuong-nhieu-khoa-thong-minh-tai-ha-noi-20230625170424680.htm?_gl=1*18oxibc*_ga*MTkxNjE3OTU1MC4xNjY4NTk3NjE3

https://www.24h.com.vn/tin-tuc-trong-ngay/nguyen-nhan-vu-xe-cho-doi-bong-da-tre-quang-nam-tai-nan-c46a1478478.html

### *Reproduce Step:
1, User A send user B a link

### **Actual:** 
A part of Dantri link is not underline

https://user-images.githubusercontent.com/124866146/248577100-003d02d3-9356-4674-8f41-57f369570d79.mp4

### **Expected:** 
All url will be underline

### PR DESCRIPTION
**Pull request by imGok**
image

This Closes #221

### CODE REVIEW

**Comment by sherlockvn** [comment]
> image
> 
> This Closes #221

Can you try with more links ?

File: lib/utils/string_extension.dart [hunk 1]
```diff
@@ -89,4 +89,34 @@ extension StringCasingExtension on String {
     final String? firstValidLink = matches.firstWhere((link) => AnyLinkPreview.isValidLink(link!));
     return firstValidLink;
   }
+
+  // Removes markdowned links from a string based on the unformatted text
+  // Workaround for content['formatted_body'] which formats urls in a way that makes them unusable
```
Comment (sherlockvn) [inline lib/utils/string_extension.dart:94]: Do you make a request change to upstream library

Comment (imGok) [reply lib/utils/string_extension.dart:94]: no, didn't really look into it

**Comment by imGok** [comment]
image

**Review by hoangdat — COMMENTED** [review]
Can you write the test to make sure it is not side effect?

File: lib/utils/string_extension.dart [hunk 2]
```diff
@@ -89,4 +89,34 @@ extension StringCasingExtension on String {
...
+      final Match unformattedLinkMatch = unformattedLinksMatches.elementAt(i);
+
+      final String formattedLink = formattedLinkMatch.group(0)!;
+      final String unformattedLink = unformattedLinkMatch.group(0)!;
+
+      unMarkdownedText = unMarkdownedText.replaceFirst(formattedLink, unformattedLink);
+    }
+
+    return unMarkdownedText;
+  }
```
Comment (nqhhdev) [inline lib/utils/string_extension.dart:121]: 1. Replace `this` keyword with the `unformattedText` parameter, as it should be used instead.
2. Change the for loop to a while loop using iterators to ensure that we iterate over both sets of matches simultaneously.
3. Change `elementAt(i)` to `current`
3. Replacet `this` with `unformattedText`  in the line where you replace the formatted link with the unformatted link.

**Review by nqhhdev — APPROVED** [review]

File: lib/utils/string_extension.dart (hunk 2 above)
Comment (imGok) [reply lib/utils/string_extension.dart:121]: 1. I want to keep formatted text / markdown on everything except URLs. IMO it's easier to start from formatted and replace only links so we keep everything else

I'll work on 2, 3

Comment (nqhhdev) [reply lib/utils/string_extension.dart:121]:
```
while (formattedIterator.moveNext() && unformattedIterator.moveNext()) {
    final Match formattedLinkMatch = formattedIterator.current;
    final Match unformattedLinkMatch = unformattedIterator.current;

    final String formattedLink = formattedLinkMatch.group(0)!;
    final String unformattedLink = unformattedLinkMatch.group(0)!;

    unMarkdownedText = unMarkdownedText.replaceFirst(formattedLink, unformattedLink);
  }
```

WDYT

Comment (sherlockvn) [reply lib/utils/string_extension.dart:121]: how about:
String unMarkdownLinks(String unformattedText) {
  final RegExp regex = RegExp(r'https:\/\/[^\s]+');

  final Iterable<Match> formattedLinksMatches = regex.allMatches(unformattedText);
  final List<String> formattedLinks = formattedLinksMatches.map((match) => match.group(0)!).toList();

  if (formattedLinks.isEmpty) {
    return unformattedText;
  }

  String unMarkdownedText = unformattedText;

  for (String formattedLink in formattedLinks) {
    final String unformattedLink = formattedLink;

    unMarkdownedText = unMarkdownedText.replaceFirst(formattedLink, unformattedLink);
  }

  return unMarkdownedText;
}

Comment (imGok) [reply lib/utils/string_extension.dart]: I'll put @nqhhdev s version, it's cleaner and safer

**Comment by imGok** [comment]
recheck after new changes : 
tests succeeded + 
image

File: test/string_extension_test.dart [hunk 3]
```diff
@@ -45,4 +45,49 @@ void main() {
 
     expect(result, isEmpty);
   });
+
+  test('unMarkdownLinks should replace links when there is only one link', () {
+    const messageWithOneLinkUnMarkdowned = "https://superman.com/this-is-a-*superman*-link";
+    const messageWithOneLinkMarkdowned = "https://superman.com/this-is-a-<em>superman</em>-link";
+
+    final result = messageWithOneLinkMarkdowned.unMarkdownLinks(messageWithOneLinkUnMarkdowned);
```
Comment (hoangdat) [inline test/string_extension_test.dart:53]: what is the input? what is the parameter? hard for me?

File: test/string_extension_test.dart [hunk 4]
```diff
@@ -45,4 +45,49 @@ void main() {
...
+  test('unMarkdownLinks should replace links when there are multiple links', () {
+    const messageWithMultipleLinksUnMarkdowned = "https://superman.com/this-is-a-*superman*-link https://batman.com/this-is-a-*batman*-link";
+    const messageWithMultipleLinksMarkdowned = "https://superman.com/this-is-a-<em>superman</em>-link https://batman.com/this-is-a-<em>batman</em>-link";
+
+    final result = messageWithMultipleLinksMarkdowned.unMarkdownLinks(messageWithMultipleLinksUnMarkdowned);
+
+    expect(result, equals(messageWithMultipleLinksUnMarkdowned));
+  });
+
+  test('unMarkdownLinks should replace links when there are links and text', () {
```
Comment (hoangdat) [inline test/string_extension_test.dart:67]: how about the case:
1. `is this serious https://superman.com/this-is-a-<em>superman</em>-link`
2. `is this serious-https://superman.com/this-is-a-<em>superman</em>-link`
3. `https://superman.com/this-is-a-<em>superman</em>-link hello guys`
4. `https://superman.com/this-is-a-<em>supe hello guys`
5. `https://superman.com/this-is-a-<em>superman</em>-link hello guys https://bat`
6. `https://superman.com/this-is-a-<em>superman</em>-link hello guys https://batman.com/this-`

File: test/string_extension_test.dart (hunk 3 above)
Comment (imGok) [reply test/string_extension_test.dart:53]: input is the message with markdown tags such as <em> etc, and parameter is the raw message with no markdown tag

Comment (hoangdat) [reply test/string_extension_test.dart:53]: how other developer can use your function easily. still hard for me to use. why we know the input when we have a messy message? or smt like that?

File: test/string_extension_test.dart (hunk 4 above)
Comment (imGok) [reply test/string_extension_test.dart:67]: for me 3,4,5,6 are already covered
I'll put a test for 1 for text + link and 2 to test the regex https

File: test/string_extension_test.dart (hunk 3 above)
Comment (hoangdat) [reply test/string_extension_test.dart:53]: @imGok maybe you miss this comment

File: test/string_extension_test.dart [hunk 5]
```diff
@@ -69,8 +69,20 @@ void main() {
...
+
+    final result2 = messageWithTextAndOneLinkMarkdowned.unMarkdownLinks(messageWithTextAndOneLinkUnMarkdowned);
+    expect(result2, equals(messageWithTextAndOneLinkUnMarkdowned));
+
+    // Link is not replaced because it is not a valid link
+    const messageWithFakeLinkUnMarkdowned = "is this serious-https://superman.com/this-is-a-<em>superman</em>-link";
+    const messageWithFakeLinkMarkdowned = "is this serious-https://superman.com/this-is-a-<em>superman</em>-link";
+
+    final result3 = messageWithFakeLinkMarkdowned.unMarkdownLinks(messageWithFakeLinkUnMarkdowned);
+    expect(result3, equals(messageWithFakeLinkUnMarkdowned));
```
Comment (hoangdat) [inline test/string_extension_test.dart:85]: please separate each case in each method

File: test/string_extension_test.dart [hunk 6]
```diff
@@ -69,8 +69,20 @@ void main() {
...
     expect(result, equals(messageWithMultipleLinksUnMarkdowned));
+
+    const messageWithTextAndOneLinkUnMarkdowned = "is this serious https://superman.com/this-is-a-*superman*-link";
+    const messageWithTextAndOneLinkMarkdowned = "is this serious https://superman.com/this-is-a-<em>superman</em>-link";
+
+    final result2 = messageWithTextAndOneLinkMarkdowned.unMarkdownLinks(messageWithTextAndOneLinkUnMarkdowned);
+    expect(result2, equals(messageWithTextAndOneLinkUnMarkdowned));
+
+    // Link is not replaced because it is not a valid link
+    const messageWithFakeLinkUnMarkdowned = "is this serious-https://superman.com/this-is-a-<em>superman</em>-link";
```
Comment (hoangdat) [inline test/string_extension_test.dart:81]: no, IMO, it is still the link, but a little bit typo. But ok for the moment, we can make it better step by step

File: test/string_extension_test.dart (hunk 4 above)
Comment (hoangdat) [reply test/string_extension_test.dart:67]: i can not find 3,4,5,6 in test-cases. please separate it in method. Thanks

Comment (imGok) [reply test/string_extension_test.dart:67]: updated

File: test/string_extension_test.dart (hunk 3 above)
Comment (imGok) [reply test/string_extension_test.dart:53]: It's used when the developer knows that the content of the message may contain links

Comment (imGok) [reply test/string_extension_test.dart:53]: this function is just to prevent links to be unusable because of markdown

Comment (hoangdat) [reply test/string_extension_test.dart:53]: ok tmr please more detail for me on it. Thanks

**Comment by hoangdat** [comment]
please demo the link + emoji

**Comment by imGok** [comment]
emojimarkdown

**Review by hoangdat — APPROVED** [review]

**Review by drminh2807 — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**